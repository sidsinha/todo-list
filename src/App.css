/* .App {
  text-align: center;
} */

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

/* .App-header {
  background-color: #282c34;
  max-height: 40vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
} */

/* .App-link {
  color: #61dafb;
} */

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}


/* .content {
  width: 1000px;
  margin: 0 auto;
  background-color: #f1f1f1;
  box-sizing: border-box;
} */
/* .todolist {
  border: 1px solid #ccc;
  width: 500px;
  margin: 0 auto;
  background-color: #ccc;
} */
/* .todo {
  background-color: #fff;
  margin: 10px;
  padding: 10px;
} */
/* span.todo-index {
  float: left;
  font-weight: bold;
} */
/* span.todo-name {
  float: left;
  margin-left: 10px;
} */

/* span.todo-action {
  float: right;
  font-size: 12px;
  color: #fff;
  cursor: pointer;
  padding: 2px 7px;
} */
/* .cancel {
  background-color: orange;
} */
/* .edit {
  background-color: yellowgreen;
} */
/* .delete {
  background-color: #ff0000;
} */
/* .content {
  width: 1000px;
  margin: 0 auto;
  background-color: #f1f1f1;
  padding: 25px 10px;
} */
/* .create-todo {
  padding: 20px;
} */
/* input[type="text"] {
  width: 300px;
  padding: 10px;
} */
/* button {
  padding: 10px 35px;
  font-weight: bold;
} */
/* .clear {
  clear: both;
} */