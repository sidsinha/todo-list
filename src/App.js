import React, { useEffect, useState, lazy, Suspense } from 'react';

import CreateTodo from "./components/create-todo";
import TodoList from "./components/todo-list";
import TodoContext from "./context/todo-context";

// const TodoList2 = lazy(() => import("./components/todo-list"));

const json = require("./data.json");

import logo from './logo.svg';
import './App.css';

const App = () => {

  function useEditObj() {
    const [editItemObj, setEditItemObj] = useState({});

    const emptyEditObj = () => {
      setEditItemObj({});
    }
    const setEditObj = (obj) => {
      setEditItemObj(obj);
    }
    return [editItemObj, emptyEditObj, setEditObj];
  }

  function useEditMode() {
    const [editMode, setEditMode] = useState(false);

    const enableEditMode = () => {
      setEditMode(true);
    }
    const disableEditMode = () => {
      setEditMode(false);
    }
    return [editMode, enableEditMode, disableEditMode];
  }
  
  const [todoData, setTodoData] = useState([]);
  const [editItemObj, emptyEditObj, setEditObj] = useEditObj();
  const [editMode, enableEditMode, disableEditMode] = useEditMode();


  useEffect(() => {
      const timer = setTimeout(()=>{ 
        setTodoData(json.data);
      }, 1000);
      return () => {
        clearTimeout(timer);
      };
  }, []);

  const addTodoFn = name => {
    const newId  = todoData.length + 1;
    const newTodoObj = {
      id: newId,
      name: name
    }
    setTodoData([
      ...todoData,
      newTodoObj
    ]);
  }
  const updateTodoFn = name => {
    const todoItem = todoData[editItemObj.index];
    todoItem.name = name;
    todoItem.isEditing = false;
    setTodoData([
      ...todoData
    ]);
    disableEditMode();
    emptyEditObj()
  }

  const removeTodoFn = id => {
    setTodoData(todoData.filter((todo)=>todo.id !== id));
    disableEditMode();
    emptyEditObj()
  }

  const editTodoFn = index => {
    const todoItem = todoData[index];
    todoItem.isEditing = true;
    setTodoData([
      ...todoData
    ]);
    enableEditMode();
    setEditObj({
      index: index,
      name: todoItem.name
    })
  }

  const cancelTodoFn = index => {
    const todoItem = todoData[index];
    todoItem.isEditing = false;
    setTodoData([
      ...todoData
    ]);
    disableEditMode();
  }

  return (
    <div className="App">
      <div className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        <h2>Welcome to Todo App</h2>
      </div>
        <div className="content">
          <h4>You have {todoData.length} todo items</h4>
          <CreateTodo 
            editItemObj={editItemObj} 
            editMode={editMode}
            addTodoFn={addTodoFn}
            updateTodoFn={updateTodoFn}
          />
          <TodoContext.Provider value={{
            cancelTodoFn: cancelTodoFn,
            editTodoFn: editTodoFn, 
            removeTodoFn: removeTodoFn,
            editMode: editMode
          }}
        >
          <TodoList todoData={todoData} />
        </TodoContext.Provider>
        </div>
    </div>
  );
}

export default App;
