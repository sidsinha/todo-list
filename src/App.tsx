import React, { useEffect, useState } from 'react';

//Importing Other react components
import CreateTodo from "./components/create-todo";
import TodoList from "./components/todo-list";

//Importing Context
import TodoContext from "./context/todo-context";

//Importing Interfaces
import { ITodo, APIResponse } from "./types";

//Importing Hooks
import useEditMode from "./hooks/useEditMode";
import useEditObj from "./hooks/useEditObj";
import useTodos from "./hooks/useTodos";

//Importing Styles, Images
import "./App.css";
import logo from "./logo.svg";

// const jsonData = require("./data.json");

const App: React.FC = () => {

  const { status, data } = useTodos();

  const [todoData, setTodoData] = useState<ITodo[]>([]);
  const [editItemObj, emptyEditObj, setEditObj] = useEditObj();
  const [editMode, enableEditMode, disableEditMode] = useEditMode();

  useEffect(() => {
    if(status === "success") {
      setTodoData(data);
    }
  }, [status, data]);

  const addTodoFn = (name:string):void => {
    const newId:number  = todoData.length + 1;
    const newTodoObj:ITodo = {
      id: newId,
      name: name,
      isEditing: false
    }
    setTodoData([
      ...todoData,
      newTodoObj
    ]);
  }

  const updateTodoFn = (name:string):void => {
    const todoItem:any = todoData[editItemObj.index];
      todoItem.name = name;
      todoItem.isEditing = false;
    
    setTodoData([
      ...todoData
    ]);
    disableEditMode();
    emptyEditObj()
  }

  const removeTodoFn = (id:number):void => {
    setTodoData(todoData.filter((todo)=>todo.id !== id));
    disableEditMode();
    emptyEditObj()
  }

  const editTodoFn = (index:number):void => {
    const todoItem = todoData[index];
    todoItem.isEditing = true;
    setTodoData([
      ...todoData
    ]);
    enableEditMode();
    setEditObj({
      index: index,
      name: todoItem.name
    })
  }

  const cancelTodoFn = (index:number):void => {
    const todoItem = todoData[index];
    todoItem.isEditing = false;
    setTodoData([
      ...todoData
    ]);
    disableEditMode();
  }

  return (
      <div className="tc courier">
        <div className="bg-black">
          <img src={logo} className="App-logo" alt="logo" />
          <h2 className="white">Welcome to Todo App</h2>
        </div>
        <div className="w-70 w-90-m w-100-s mt0 center bg-light-gray pa3">
          <h4>You have {todoData.length} todo items</h4>
          <CreateTodo 
            editItemObj={editItemObj} 
            editMode={editMode}
            addTodoFn={addTodoFn}
            updateTodoFn={updateTodoFn}
          />
          <TodoContext.Provider value={{
            cancelTodoFn: cancelTodoFn,
            editTodoFn: editTodoFn, 
            removeTodoFn: removeTodoFn,
            editMode: editMode
          }}
        >
          <TodoList todoData={todoData} />
        </TodoContext.Provider>
        </div>
      </div>
  );
}

export default App;
