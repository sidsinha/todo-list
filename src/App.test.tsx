import React from 'react';
import { render } from '@testing-library/react';
import App from './App';

import { setupWorker, graphql } from 'msw'


const fetchDataFromGraphQL = async() => {
  const query = `
      query {
      posts {
          id
          name
          isEditing
      }
      }
  `;
  const url = "https://api-us-west-2.graphcms.com/v2/ckb6w8kn5054n01wk0w77cc0k/master";
  const opts = {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ query })
  };
  await new Promise(resolve => setTimeout(resolve, 2000));
  const response = await fetch(url, opts);
  const data = await response.json();
  return data?.data?.posts;
}

const worker = setupWorker(
  graphql.query(
    'GetUserDetail',
    (req, res, ctx) => {
      const { userId } = req.variables
      
      return res(
        ctx.data({
          user: {
            firstName: 'John',
            lastName: 'Maverick'
          }
        })
      )
    })
)

beforeAll(() => worker.start())

test('renders Todo App', async() => {
    const { container } = render(<App />)
    expect(container.firstChild).toMatchSnapshot();
  // expect(getByText('Welcome to Todo App')).toBeInTheDocument()
  });
