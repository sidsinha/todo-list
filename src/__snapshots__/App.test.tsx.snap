// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Todo App 1`] = `
<div
  class="tc courier"
>
  <div
    class="bg-black"
  >
    <img
      alt="logo"
      class="App-logo"
      src="logo.svg"
    />
    <h2
      class="white"
    >
      Welcome to Todo App
    </h2>
  </div>
  <div
    class="w-70 w-90-m w-100-s mt0 center bg-light-gray pa3"
  >
    <h4>
      You have 
      0
       todo items
    </h4>
    <div
      class="pa3"
    >
      <input
        class="pa2 w-40 w-60-m"
        name="todo_name"
        placeholder="Enter new todo item"
        type="text"
      />
      <button
        class="pv2 ph4"
        type="submit"
      >
        ADD
      </button>
    </div>
    <div
      class="ba b--silver w-70 w-100-m center bg-silver"
    />
  </div>
</div>
`;
